connectivity/perf/benchmarks/netperf/perfpod.go

```diff
@@ -0,0 +1,164 @@
+					"TCP_RR",
+					"TCP_STREAM",
+					"UDP_RR",
+					"UDP_STREAM",
+					"TCP_CRR",
+				}
+
+				for _, test := range tests {
+					action := t.NewAction(s, tool, &c, server, features.IPFamilyV4)
+					action.CollectFlows = false
+					action.Run(func(a *check.Action) {
+						k := common.PerfTests{
+							Test:     test,
+							Tool:     tool,
+							SameNode: sameNode,
+							Sample:   sample,
+							Duration: duration,
+							Scenario: scenarioName,
+						}
+						perfResult := netperf(ctx, server.Pod.Status.PodIP, k, a)
+						t.Context().PerfResults = append(t.Context().PerfResults, common.PerfSummary{PerfTest: k, Result: perfResult})
+
+					})
+				}
+			}
+		}
+	}
+}
+
+func buildExecCommand(test string, sip string, duration time.Duration, args []string) []string {
+	exec := []string{"/usr/local/bin/netperf", "-H", sip, "-l", duration.String(), "-t", test, "--", "-R", "1", "-m", fmt.Sprintf("%d", messageSize)}
+	exec = append(exec, args...)
+
+	return exec
+}
+
+func parseDuration(a *check.Action, value string) time.Duration {
+	res, err := time.ParseDuration(value + "us") // by default latencies in netperf are reported in microseconds
+	if err != nil {
+		a.Fatalf("Unable to process netperf result, duration: %s", value)
+	}
+	return res
+}
+
+func parseFloat(a *check.Action, value string) float64 {
+	res, err := strconv.ParseFloat(value, 64)
+	if err != nil {
+		a.Fatalf("Unable to process netperf result, float: %s", value)
+	}
+	return res
+}
+
+func netperf(ctx context.Context, sip string, perfTest common.PerfTests, a *check.Action) common.PerfResult {
+	args := []string{"-o", "MIN_LATENCY,MEAN_LATENCY,MAX_LATENCY,P50_LATENCY,P90_LATENCY,P99_LATENCY,TRANSACTION_RATE,THROUGHPUT,THROUGHPUT_UNITS"}
+	exec := buildExecCommand(perfTest.Test, sip, perfTest.Duration, args)
+
+	a.ExecInPod(ctx, exec)
+	output := a.CmdOutput()
+	a.Debugf("Netperf output: ", output)
+	lines := strings.Split(output, "\n")
+	resultsLine := lines[len(lines)-2]
+	values := strings.Split(resultsLine, ",")
+	a.Debugf("Numbers: %v", values)
+
+	res := common.PerfResult{
+		Timestamp: time.Now(),
+		Latency: &common.LatencyMetric{
+			Min:    parseDuration(a, values[0]),
+			Avg:    parseDuration(a, values[1]),
+			Max:    parseDuration(a, values[2]),
+			Perc50: parseDuration(a, values[3]),
+			Perc90: parseDuration(a, values[4]),
+			Perc99: parseDuration(a, values[5]),
+		},
+		TransactionRateMetric: &common.TransactionRateMetric{
+			TransactionRate: parseFloat(a, values[6]),
+		},
+		ThroughputMetric: &common.ThroughputMetric{
+			Throughput: parseFloat(a, values[7]) * 1000000, // by default throughput has unit "10^6bits/s", we verify that later
+		},
+	}
+
+	if strings.HasSuffix(perfTest.Test, "_STREAM") {
+		// We don't want to report transaction rate or latency
+		res.TransactionRateMetric = nil
+		res.Latency = nil
+		// Verify that throughput unit is 10^6bits/s
+		if values[8] != "10^6bits/s" {
+			a.Fatal("Unable to process netperf result")
```

Comment:
Perhaps this call should return an err and let the call site decide vs bombing?

Reply:
I don't think it's worth it - we already handle a bunch of errors through `a.Fatal` - FYI this is just failing a single action, not the whole program.